server/routerlicious/packages/services-shared/src/healthCheckEndpoints.ts

```diff
@@ -0,0 +1,102 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Router } from "express";
+import { StartupChecker } from "./startupChecker";
+import { Lumberjack } from "@fluidframework/server-services-telemetry";
+
+/**
+ * Readiness status of a service or functionality.
+ * @internal
+ */
+export interface IReadinessStatus {
+	/**
+	 * Whether the service/functionality is ready for use.
+	 */
+	ready: boolean;
+
+	/**
+	 * Optional exception if an error occurs.
+	 */
+	exception?: any;
+}
+
+/**
+ * Checks if a service or functionality is ready for use.
+ * @internal
+ */
+export interface IReadinessCheck {
+	/**
+	 * Whether the service/functionality is ready for use.
+	 */
+	isReady(): Promise<IReadinessStatus>;
```

Comment:
Ideally, both readiness and liveness APIs would be evaluated based on signals originating from various places within core business logic of the service functionality they serve for. As such it might make more sense to design them in a way to be able to channel those signals from various places into the readiness/liveness evaluation function. How a sample implementation of isReady() is going to look like ?

Reply:
You're correct, readiness probes specifically can be used to test API readiness and connectivity to external dependencies. These signals would need to be added and injected incrementally into OSS.

We will need to implement `isReady()` for all our services.  A very basic implementation of `isReady()` could simply check connectivity to external resources for a service such as MongoDB, Redis, Storage or Kafka. In case of failures of these, we could possibly fail the readiness probe. The probe could be made even more complex to test APIs based on connectivity metrics. Without these APIs the collab experience would break.

Reply:
I was wondering how those signals from different parts of service are going to be channeled into the isReady() function for evaluation? Probably, you would need to utilize a singleton approach similar to startupChecker so that to be able to toggle the readiness state from various places in the service while exposing that same state in the readiness API.

Reply:
We will set up these IReadinessCheck instances with internal implementations and pass them in as needed. This follows the pattern used everywhere in R11s, called "dependency injection." The singleton approach specifically _does not_ allow customizing the implementation.